src/main/java/com/ibm/cloud/sdk/core/service/model/DynamicModel.java

```diff
@@ -1,14 +1,103 @@
 package com.ibm.cloud.sdk.core.service.model;
 
 import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
 
+import com.google.gson.reflect.TypeToken;
 import com.ibm.cloud.sdk.core.util.GsonSingleton;
 
 /**
- * Abstract model class for objects which may have dynamic properties attached to them,
- * which is represented with an internal map.
+ * Abstract model class for a model which supports dynamic (arbitrary) properties of type T.
  */
-public abstract class DynamicModel extends HashMap<String, Object> implements ObjectModel {
+public abstract class DynamicModel<T> implements ObjectModel {
+  private TypeToken<T> additionalPropertyTypeToken;
+
+  // The set of dynamic properties associated with this object.
+  private Map<String, T> dynamicProperties = new HashMap<>();
+
+  /**
+   * Force use of 1-arg ctor.
+   */
+  @SuppressWarnings("unused")
+  private DynamicModel() {
+  }
+
+  /**
+   * This ctor accepts a TypeToken instance that represents the type of values stored in the map.
+   *
+   * @param t
+   *          the TypeToken which represents the type of map values
+   */
+  public DynamicModel(TypeToken<T> t) {
+    this.additionalPropertyTypeToken = t;
+  }
+
+  /**
+   * Returns the TypeToken which describes the type of additional properties stored in the map.
+   * @return The TypeToken which descriptes the map value type
+   */
+  public TypeToken<T> getAdditionalPropertyTypeToken() {
+    return this.additionalPropertyTypeToken;
+  }
+
+  /**
+   * Sets an arbitrary property.
+   *
+   * @param key
+   *          the name of the property to set
+   * @param value
+   *          the value of the property to be set
+   * @return the previous value of the property, or null if the property was not previously set
```

Comment:
Yeah, the reason I followed the "Map.put" model is so that I could easily tell if we have a duplicate property while deserializing a JSON string (by checking to see if the result of the set operation is non-null).   Technically, I don't need to do it this way... i could first check to see if a value already exists for a given key, but the Map "put" method (which is really just wrapped by DynamicModel.setProperty) does that all in one step so thought it would be better to just follow that same pattern with the setProperty method.